lib/vnet/vnet.go

```diff
@@ -339,6 +395,122 @@ func (m *Manager) statsHandler(ctx context.Context) error {
 }
 
+func (m *Manager) handleUDP(req *udp.ForwarderRequest) {
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		m.handleUDPConcurrent(req)
+	}()
+}
+
+func (m *Manager) handleUDPConcurrent(req *udp.ForwarderRequest) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	id := req.ID()
+	slog := m.slog.With("request", id)
+	slog.DebugContext(ctx, "Handling UDP request.")
+	defer slog.DebugContext(ctx, "Finished handling UDP request.")
+
+	handler, ok := m.getUDPHandler(id.LocalAddress)
+	if !ok {
+		slog.DebugContext(ctx, "No handler for address.", "addr", id.LocalAddress)
+		return
+	}
+
+	var wq waiter.Queue
+	endpoint, err := req.CreateEndpoint(&wq)
+	if err != nil {
+		slog.With("error", err).ErrorContext(ctx, "Failed to create UDP endpoint.")
+		return
+	}
+
+	conn := gonet.NewUDPConn(m.stack, &wq, endpoint)
+	defer conn.Close()
+
+	if err := handler.HandleUDP(ctx, conn); err != nil {
+		slog.DebugContext(ctx, "Error handling UDP conn.", "error", err)
+	}
+}
+
+func (m *Manager) getUDPHandler(addr tcpip.Address) (UDPHandler, bool) {
+	m.state.mu.RLock()
+	defer m.state.mu.RUnlock()
+	handler, ok := m.state.udpHandlers[addr]
+	return handler, ok
+}
+
+func (m *Manager) assignUDPHandler(addr tcpip.Address, handler UDPHandler) error {
+	m.state.mu.Lock()
+	defer m.state.mu.Unlock()
+	if _, ok := m.state.udpHandlers[addr]; ok {
+		return trace.AlreadyExists("Handler for %s is already set", addr)
+	}
+	if err := m.addProtocolAddress(addr); err != nil {
+		return trace.Wrap(err)
+	}
+	m.state.udpHandlers[addr] = handler
+	return nil
+}
+
+// ResolveA implements [dns.Resolver.ResolveA].
+func (m *Manager) ResolveA(ctx context.Context, fqdn string) (dns.Result, error) {
+	if _, ok := m.appIPv6(fqdn); ok {
+		// Matched a known app but not supporting IPv4 yet, return NoRecord.
+		return dns.Result{
+			NoRecord: true,
+		}, nil
+	}
+
+	_, found, err := m.tcpHandlerResolver.ResolveTCPHandler(ctx, fqdn)
+	if err != nil {
+		return dns.Result{}, trace.Wrap(err, "resolving TCP handler for fqdn %q", fqdn)
+	}
+	if !found {
+		// Did not find any known app, forward the DNS request upstream.
+		return dns.Result{}, nil
+	}
+
+	// Matched a known app but not supporting IPv4 yet, return NoRecord.
+	return dns.Result{
+		NoRecord: true,
+	}, nil
+}
+
+// ResolveAAAA implements [dns.Resolver.ResolveAAAA].
+func (m *Manager) ResolveAAAA(ctx context.Context, fqdn string) (dns.Result, error) {
+	if ip, ok := m.appIPv6(fqdn); ok {
+		return dns.Result{
+			AAAA: ip.As16(),
+		}, nil
+	}
+
+	tcpHandler, found, err := m.tcpHandlerResolver.ResolveTCPHandler(ctx, fqdn)
+	if err != nil {
+		return dns.Result{}, trace.Wrap(err, "resolving TCP handler for fqdn %q", fqdn)
+	}
+	if !found {
+		// Did not find any known app, forward the DNS request upstream.
+		return dns.Result{}, nil
+	}
+
+	addr, err := m.assignTCPHandler(tcpHandler, fqdn)
+	if err != nil {
+		return dns.Result{}, trace.Wrap(err, "assigning address to handler for %q", fqdn)
+	}
```

Comment:
Good point, you know what, I should probably do the `singleflight` at this layer instead of in app_resolver.go and it should include assigning the IP to the handler. I'll go ahead and make that change

Reply:
doing this in the latest